openedx/core/djangoapps/content/search/api.py

```diff
@@ -0,0 +1,328 @@
+        # Mark which attributes can be used for filtering/faceted search:
+        client.index(temp_index_name).update_filterable_attributes([
+            Fields.block_type,
+            Fields.context_key,
+            Fields.org,
+            Fields.tags,
+            Fields.type,
+        ])
+
+        ############## Libraries ##############
+        status_cb("Indexing libraries...")
+        for lib_key in lib_keys:
+            status_cb(f"{num_contexts_done + 1}/{num_contexts}. Now indexing library {lib_key}")
+            docs = []
+            for component in lib_api.get_library_components(lib_key):
+                metadata = lib_api.LibraryXBlockMetadata.from_component(lib_key, component)
+                doc = searchable_doc_for_library_block(metadata)
+                docs.append(doc)
+                num_blocks_done += 1
+            # Add all the docs in this library at once (usually faster than adding one at a time):
+            _wait_for_meili_task(client.index(temp_index_name).add_documents(docs))
+            num_contexts_done += 1
+
+        ############## Courses ##############
+        status_cb("Indexing courses...")
+        for course in all_courses:
+            status_cb(
+                f"{num_contexts_done + 1}/{num_contexts}. Now indexing course {course.display_name} ({course.id})"
+            )
+            docs = []
+
+            def add_with_children(block):
+                """ Recursively index the given XBlock/component """
+                doc = searchable_doc_for_course_block(block)
+                docs.append(doc)  # pylint: disable=cell-var-from-loop
+                _recurse_children(block, add_with_children)  # pylint: disable=cell-var-from-loop
+
+            _recurse_children(course, add_with_children)
+
+            # Add all the docs in this course at once (usually faster than adding one at a time):
+            _wait_for_meili_task(client.index(temp_index_name).add_documents(docs))
+            num_contexts_done += 1
+            num_blocks_done += len(docs)
+
+    status_cb(f"Done! {num_blocks_done} blocks indexed across {num_contexts_done} courses and libraries.")
+
+
+def upsert_xblock_index_doc(
+    usage_key: UsageKey, recursive: bool = True, update_metadata: bool = True, update_tags: bool = True
+) -> None:
+    """
+    Creates or updates the document for the given XBlock in the search index
+
+    Args:
+        usage_key (UsageKey): The usage key of the XBlock to index
+        recursive (bool): If True, also index all children of the XBlock
+        update_metadata (bool): If True, update the metadata of the XBlock
+        update_tags (bool): If True, update the tags of the XBlock
+    """
+    # If there is a rebuild in progress, wait for it to finish
+    # This could still be a problem if a rebuild starts right after this check, but we don't want to prevent
+    # concurrent updates entirely.
+    _wait_index_rebuild_lock(INDEX_NAME)
```

Comment:
What do you think about an alternative approach: if a rebuild is in progress, do the upsert in _both_ the old index and the new index. Then you don't need to use locks or worry as much about race conditions.

Reply:
Hmm. The problem is how to safely check that a rebuild is in progress and update atomically without a lock. Could you help me if you see a way?

We could accept some race conditions, and, thinking about it a bit more, the current scenario is fine: if the command is called AFTER the XBLOCK event, the new index will have the correct data (even if it discards the updates from the event because the XBLOCK is already updated).

Or we could stop replacing the index and accept the performance hit to simplify the code.

What do you think @bradenmacdonald ?

Reply:
I like the way that you're tracking whether an index update is occurring using the cache. You can stick with that but instead of making it a boolean you can store the name of the new index in that "lock" while the rebuild is happening. Then the normal update code can detect that, and apply the changes to both indexes.

I don't think anything has to be atomic? As long as you make the update in both indexes, it will be fine. Even if the "reindex" command later overwrites it, it will still be the latest version.

Reply:
Done: d63195dfbdc7d93ce8f39abe364b413a3897b649